TabnineSelfHosted/src/main/java/com/tabnineSelfHosted/statusBar/TabnineSelfHostedStatusBarWidget.kt

```diff
@@ -0,0 +1,111 @@
+package com.tabnineSelfHosted.statusBar
+
+import com.intellij.ide.DataManager
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.popup.JBPopupFactory
+import com.intellij.openapi.ui.popup.ListPopup
+import com.intellij.openapi.wm.StatusBarWidget
+import com.intellij.openapi.wm.StatusBarWidget.MultipleTextValuesPresentation
+import com.intellij.openapi.wm.impl.status.EditorBasedWidget
+import com.intellij.util.Consumer
+import com.tabnineCommon.binary.requests.config.CloudConnectionHealthStatus
+import com.tabnineCommon.binary.requests.config.StateResponse
+import com.tabnineCommon.general.StaticConfig
+import com.tabnineCommon.lifecycle.BinaryStateChangeNotifier
+import com.tabnineSelfHosted.statusBar.SelfHostedStatusBarActions.buildStatusBarActionsGroup
+import java.awt.event.MouseEvent
+import javax.swing.Icon
+
+class TabnineSelfHostedStatusBarWidget(project: Project) : EditorBasedWidget(project), StatusBarWidget, MultipleTextValuesPresentation {
+    private var cloudConnectionHealthStatus = CloudConnectionHealthStatus.Ok
+    private var username: String? = ""
+
+    init {
+        // register for state changes (we will get notified whenever the state changes)
+        ApplicationManager.getApplication()
+            .messageBus
+            .connect(this)
+            .subscribe(
+                BinaryStateChangeNotifier.STATE_CHANGED_TOPIC,
+                BinaryStateChangeNotifier { (_, _, _, cloudConnectionHealthStatus1, _, userName): StateResponse ->
+                    cloudConnectionHealthStatus = cloudConnectionHealthStatus1
+                    username = userName
+                    update()
+                }
+            )
+    }
+
+    override fun getIcon(): Icon? {
```

Comment:
consider extracting some of the logic in the functions below to separate file/s for better readability of the files